Add a cargo deny workflow

kayabaNerve opened the pull request:
Also trims out a pointless submodule checkout (we have none).

Relevant to #81 and #82.

kayabaNerve committed:
Add a cargo deny workflow

Also trims out a pointless submodule checkout (we have none).

kayabaNerve committed:
Remove no longer relevant advisories/allowances

kayabaNerve committed:
Patch for array-bytes

kayabaNerve committed:
Remove unused properties

kayabaNerve committed:
Restore chrono advisory

kayabaNerve committed:
Allow MPL-2.0, correct GPL-3.0 allowance specification

kayabaNerve committed:
Properly ban copyleft, run on all crates

kayabaNerve committed:
Exceptions for Serai crates (AGPL-3.0)

kayabaNerve committed:
Remove top comments

kayabaNerve committed:
Clarify reasoning for not checking advisories in CI

TheArchitect108 commented on deny.toml:
```diff
@@ -0,0 +1,73 @@
+db-path = "~/.cargo/advisory-db"
+db-urls = ["https://github.com/rustsec/advisory-db"]
+
+vulnerability = "deny"
```
We should add a security threshold so this doesn't trigger on every little thing. Medium?

Threshold for security vulnerabilities, any vulnerability with a CVSS score lower than the range specified will be ignored. Note that ignored advisories will still output a note when they are encountered.
* None - CVSS Score 0.0
* Low - CVSS Score 0.1 - 3.9
* Medium - CVSS Score 4.0 - 6.9
* High - CVSS Score 7.0 - 8.9
* Critical - CVSS Score 9.0 - 10.0
`severity-threshold = `

kayabaNerve replied:
I'm not against this but I'd rather have the heads up so we can review it ourselves and then make a new commit explicitly allowing it. I will note the CI doesn't even run advisory checks right now for this reason, but I currently lean towards having it check everything and doing the aforementioned solution, having given it a second thought. Will reach out off GitHub to discuss.

TheArchitect108 commented on deny.toml:
```diff
@@ -0,0 +1,73 @@
+]
+
+copyleft = "deny"
+allow-osi-fsf-free = "neither"
```
did you simply deny the categories because you provide an explicit list?

kayabaNerve replied:
Correct. If this is warn/allow, then they pass without erroring. They should not pass.

TheArchitect108 commented on deny.toml:
```diff
@@ -0,0 +1,73 @@
+  "RUSTSEC-2020-0071", # https://github.com/chronotope/chrono/issues/602
+]
+
+[licenses]
```
do we want `private = { ignore = true }`?

kayabaNerve replied:
No. That skips the entire tree for every unpublished crate. That means we wouldn't review our crates, nor Substrate, nor Substrate's depends.

TheArchitect108 commented on deny.toml:
```diff
@@ -0,0 +1,73 @@
+
+exceptions = [
+  { allow = ["AGPL-3.0"], name = "ethereum-serai" },
+  { allow = ["AGPL-3.0"], name = "serai-processor" },
```
I think the `private = { ignore = true }` would allow us to skip all this?

TheArchitect108 reviewed (commented):
I think we need to ADD this TOML in the dockerfile for Serai so that it copies the deny and for other services as well. Maybe concern here is only running the action workflow, but we want it to trigger for local builds too.
ADD deny.toml .

TheArchitect108 commented:
Also this does not do deep scanning and relies on the author to be honest about licensing. If the lib uses embedded libs, we also rely on the author to do the appropriate thing for supplying the license. May not be relevant at this point, but if we want deeper scanning in the future we should look at https://github.com/EmbarkStudios/cargo-about.

kayabaNerve commented:
Placing this in the Dockerfile isn't helpful, unless the Dockerfile then runs `cargo deny`. I don't feel a need to have the Dockerfile run `cargo deny` just as it shouldn't run `cargo test`.

We have clarifications available in cargo deny, and there's some level of LICENSE detection, but you're right it can misrepresented. I'm not sure `cargo about` does anything deeper. It seems to just generate a complete report/citation list.

kayabaNerve committed:
Run all checks in CI

While this may bring down an unrelated commit, we can manually review, before creating a followup commit allowing it. If it's critical, then this did its job.
